main.nf

```diff
@@ -0,0 +1,533 @@
+    section_href: 'https://github.com/nf-core/deepvariant'
+    plot_type: 'html'
+    data: |
+        <dl class=\"dl-horizontal\">
+${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }.join("\n")}
+        </dl>
+    """.stripIndent()
+
+   return yaml_file
+}
+
+
+/********************************************************************
+  process preprocessFASTA
+  Collects all the files related to the reference genome, like
+  .fai,.gz ...
+  If the user gives them as an input, they are used
+  If not they are produced in this process given only the fasta file.
+********************************************************************/
+
+if(!fai) {
+  process preprocessFAI {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaCh1
+
+      output:
+      file("${fasta}.fai") into faiCh
+
+      script:
+      """
+      samtools faidx $fasta
+      """
+  }
+}
+
+if(!fastagz) {
+  process preprocessFASTGZ {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaCh2
+
+      output:
+      file("*.gz") into (tmpFastaGzCh, fastaGzCh)
+
+      script:
+      """
+      bgzip -c ${fasta} > ${fasta}.gz
+      """
+  }
+}
+
+if(!gzfai) {
+  process preprocessGZFAI {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaCh3
+    file(fastagz) from tmpFastaGzCh
+
+    output:
+    file("*.gz.fai") into gzFaiCh
+
+    script:
+    """
+    samtools faidx $fastagz
+    """
+  }
+}
+
+if(!gzi){
+  process preprocessGZI {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaCh4
+
+    output:
+    file("*.gz.gzi") into gziCh
+
+    script:
+    """
+    bgzip -c -i ${fasta} > ${fasta}.gz
+    """
+  }
+}
+
+//channel to collect all necessary reference files
+fastaChannel = Channel.from(fastaCh).mix(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh).collect()
+
+/********************************************************************
+  process preprocessBAM
+  Takes care of the read group line.
+********************************************************************/
+
+process preprocessBAM{
+
+  tag "${bam[0]}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  set val(prefix), file(bam) from bamChannel
+
+  output:
+  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel
+
+  script:
+  """
+  mkdir ready
+  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { picard AddOrReplaceReadGroups \
+  I=${bam[0]} \
+  O=ready/${bam[0]} \
+  RGID=${params.rgid} \
+  RGLB=${params.rglb} \
+  RGPL=${params.rgpl} \
+  RGPU=${params.rgpu} \
+  RGSM=${params.rgsm};}
+  cd ready ;samtools index ${bam[0]};
+  """
+}
+
+
+//make tupule of reference files using fasta file as identifier
+fastaChannel.map{file -> tuple (1,file[0],file[1],file[2],file[3],file[4], file[5])}
```

Comment:
Typically, when remapping, I prefer to explicitly redefine all items, it's much more readable.
So in this case, I would do something like:
```
fastaChannel.map{
    fasta, bed, fai, fastaGz, faiGz, gzi ->
    [tuple (1, fasta, bed, fai, fastaGz, faiGz, gzi)]
}
```